### PR TITLE
Add a lint for NeedsPermission and OnShowRationale

### PR DESCRIPTION
**Pull request by shiraji**
see #59 and its [comment](https://github.com/hotchemi/PermissionsDispatcher/issues/59#issuecomment-181818458)

### CODE REVIEW

File: lint/src/main/java/permissions/dispatcher/NoMatchingNeedsPermissionDetector.java [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+import lombok.ast.AstVisitor;
+import lombok.ast.CompilationUnit;
+import lombok.ast.ForwardingAstVisitor;
+import lombok.ast.Node;
+
+public class NoMatchingNeedsPermissionDetector extends Detector implements Detector.JavaScanner {
+
+    public static final Issue ISSUE = Issue.create("NoMatchingNeedsPermission",
+            "No matching @NeedsPermission found",
+            "The processor would ignore @OnShowRationale annotation if there is no matching @NeedsPermissions. For instance, if there is @OnShowRationale(Manifest.permission.CAMERA), then there must be @NeedsPermission(Manifest.permission.CAMERA) in the same class.",
```
Comment (shiraji) [inline lint/src/main/java/permissions/dispatcher/NoMatchingNeedsPermissionDetector.java]: For this short and long description is not good enough for this lint. I'm not sure how to express this lint...

File: lint/src/main/java/permissions/dispatcher/NoMatchingNeedsPermissionDetector.java [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+            Category.CORRECTNESS,
+            4,
+            Severity.ERROR,
```
Comment (shiraji) [inline lint/src/main/java/permissions/dispatcher/NoMatchingNeedsPermissionDetector.java]: To be honest, I don't know how to determine this lint is error or warning...

**Comment by hotchemi** [comment]
Nice!

File: lint/src/main/java/permissions/dispatcher/NoMatchingNeedsPermissionDetector.java (hunk 1 above)
Comment (shiraji) [inline lint/src/main/java/permissions/dispatcher/NoMatchingNeedsPermissionDetector.java]: `s/matching/corresponding/` makes sense to me.

**Comment by hotchemi** [comment]
Oh, sorry I didn't check the design but in this case, it causes compile error right?

File: lint/src/main/java/permissions/dispatcher/NoCorrespondingNeedsPermissionDetector.java [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+            new Implementation(NoCorrespondingNeedsPermissionDetector.class, EnumSet.of(Scope.JAVA_FILE)));
+
+    @Override
+    public AstVisitor createJavaVisitor(JavaContext context) {
+        return new AnnotationChecker(context);
+    }
+
+    static class AnnotationChecker extends ForwardingAstVisitor {
+        private final JavaContext context;
+        private Set<String> needsPermissionName;
```
Comment (hotchemi) [inline lint/src/main/java/permissions/dispatcher/NoCorrespondingNeedsPermissionDetector.java]: I guess `needsPermissionName` and `onShowRationaleName` can be initialized like below.

```java
private static final Set<String> NEEDS_PERMISSION_NAME = new HashSet<>() {{;
  add("OnShowRationale");
  add("permissions.dispatcher.OnShowRationale");
}};


```

File: lint/src/main/java/permissions/dispatcher/NoCorrespondingNeedsPermissionDetector.java [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+
+            super.afterVisitCompilationUnit(node);
+        }
+
+        private boolean hasSameNodes(List<Node> first, List<Node> second) {
+            if (first.size() != second.size()) {
+                return false;
+            }
+
+            for (int i = 0; i < first.size(); i++) {
```
Comment (hotchemi) [inline lint/src/main/java/permissions/dispatcher/NoCorrespondingNeedsPermissionDetector.java]: Below way is good for performance I guess.
```java
for (int i = 0, size = first.size(); i < size; i++) {
```

**Comment by hotchemi** [comment]
Thx! one of the my concern is https://github.com/hotchemi/PermissionsDispatcher/pull/104#issuecomment-189031201 but other than that it's almost lgtm!

One question, How could we test it?

**Comment by shiraji** [comment]
> Oh, sorry I didn't check the design but in this case, it causes compile error right?

No. PermissionsDispatcher does not generate onShowRationale method call if the `@OnShowRatinale` does not have corresponding `@NeedsPermission`

**Comment by hotchemi** [comment]
Ah ic.

**Comment by shiraji** [comment]
Let me squash commits after your review.

**Comment by shiraji** [comment]
> One question, How could we test it?

I'm sorry. I accidentally ignore this one.

In terms of testing, since `LintDetectorTest` is not good enough for testing, yet, I don't think we can write unit testing for Lint at this moment. However, we can provide sample module that shows all lint warnings.

**Comment by aurae** [comment]
`message`
*Useless @OnShowRationale declaration*

`summary`
*The method annotated with @OnShowRationale has no corresponding @NeedsPermission method, and will therefore be ignored by PermissionsDispatcher*

`explanation`
*The @OnShowRationale method with a certain signature is internally connected to another method annotated with @NeedsPermission and the same annotation value. Please ensure that there is a @NeedsPermission method with matching annotation values for this method.*

**Comment by shiraji** [comment]
Nice! I will change wording.

**Comment by hotchemi** [comment]
LInt master and English master are there:D

**Comment by hotchemi** [comment]
Thx!

**Pull request merged**